### PR TITLE
Only include stable lints in `rustdoc::all` group

### LINKED ISSUE #106289: `rustdoc::all` triggers `unknown_lints`
### Current behavior
Currently, in both stable (`1.66.0`) and nightly (`2381cbdb4 2022-12-23`), using `rustdoc::all` will trigger `unknown_lints`:
```
warning: unknown lint: `rustdoc::missing_doc_code_examples`
 --> src/lib.rs:5:18
  |
5 | #![cfg_attr(doc, warn(rustdoc::all))]
  |                  ^^^^^^^^^^^^^^^^^^
  |
  = note: the `rustdoc::missing_doc_code_examples` lint is unstable
  = note: see issue #101730 <https://github.com/rust-lang/rust/issues/101730> for more information
  = note: `#[warn(unknown_lints)]` on by default
```

### Expected behavior

The expected behavior would be that a stable lint group shouldn't use unstable lints unless the unstable feature is active.

### Problem
The current behavior is weird because you can't make the warning go away without using `allow(unknown_lints)` or enabling the unstable feature and using `allow(rustdoc::missing_doc_code_examples)`.

This broke CIs on multiple occasions that have simply been using `RUSTDOCFLAGS=-Dwarnings cargo doc`.

### Current Workaround
Currently the workaround fix is: `#![cfg_attr(doc, allow(unknown_lints), warn(rustdoc::all))]`.

### Tested Versions
Stable:
```
cargo 1.66.0 (d65d197ad 2022-11-15)
release: 1.66.0
commit-hash: d65d197ad5c6c09234369f219f943e291d4f04b9
commit-date: 2022-11-15
host: x86_64-unknown-linux-gnu
```

Nightly:
```
cargo 1.68.0-nightly (2381cbdb4 2022-12-23)
release: 1.68.0-nightly
commit-hash: 2381cbdb4e9b07090f552d34a44a529b6e620e44
commit-date: 2022-12-23
host: x86_64-unknown-linux-gnu
```

### PR DESCRIPTION
**Pull request by camelid**
Fixes #106289.

Including unstable lints in the lint group produces unintuitive behavior
on stable (see #106289). Meanwhile, if we only included unstable lints
on nightly and not on stable, we could end up with confusing bugs that
were hard to compare across versions of Rust that lacked code changes.

I think that only including stable lints in `rustdoc::all`, no matter
the release channel, is the most intuitive option. Users can then
control unstable lints individually, which is reasonable since they have
to enable the feature gates individually anyway.

r? @GuillaumeGomez

### CODE REVIEW

**Comment by camelid** [comment]
I don't believe this PR requires an FCP since it should only significantly affect unstable code. The only significant change to stable code will be removing the need to allow `unknown_lint`.

File: src/test/rustdoc-ui/check-fail.stderr
Comment (camelid) [inline src/test/rustdoc-ui/check-fail.stderr:39]: Not sure why this changed.

Comment (GuillaumeGomez) [reply src/test/rustdoc-ui/check-fail.stderr:39]: Want to investigate a bit further before approval?

Comment (camelid) [reply src/test/rustdoc-ui/check-fail.stderr:39]: Ah, I just figured it out: This error is from a different (stable) lint than the unstable `rustdoc::missing_doc_code_examples`. Because there are now two `deny`s (one for `rustdoc::all` and one for `rustdoc::missing_doc_code_examples`), the original lint level definition note is no longer for `rustdoc::all`, so it shows up again here.

So this makes sense now 👍

**Comment by camelid** [comment]
Just added an explanatory comment to the new test.

File: src/test/rustdoc-ui/check-fail.stderr
Comment (GuillaumeGomez) [reply src/test/rustdoc-ui/check-fail.stderr:39]: I see, thanks for the explanations!

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks! Considering it's changing the default behaviour of rustdoc, I think an FCP would be a good idea to answer everyone is ok with it.

@rfcbot fcp merge

**Comment by Nemo157** [comment]
My expectation would be that `rustdoc::all` would activate all enabled lints, including all stable ones and all unstable ones that have their feature active. I'm guessing that the group gets defined before parsing features so that wouldn't be possible though?

**Comment by camelid** [comment]
I see your point, but in addition to the implementation difficulty, it seems unintuitive to me for a lint to be turned on just by adding a feature gate. I personally think it's better for users to explicitly control unstable lints.

**Comment by GuillaumeGomez** [comment]
@camelid Can you rebase whenever you have time please? Then r=me.

**Pull request merged**
